test/unit/annotation_spec.js

```diff
@@ -937,29 +943,40 @@ describe('annotation', function() {
       var flags = AnnotationFieldFlag.COMB + AnnotationFieldFlag.MULTILINE +
                   AnnotationFieldFlag.PASSWORD + AnnotationFieldFlag.FILESELECT;
 
-      for (var i = 0, ii = invalidFieldFlags.length; i <= ii; i++) {
+      var xref;
```

Comment:
I don't really understand why this test has to be changed other than making it asynchronous. Why the additional `AnnotationFactory.create`?

Reply:
I'm not 100% positive I understand this unit test, but here is a deal. With synchronous execution it runs for combobox flag + invalid flags and check the condition expect(data.comb).toEqual(false) and then for the combobox only with condition  expect(data.comb).toEqual(true). In this mode one can have the same loop and calculate true/false parameter. With asynchronous model I was forced to split loop into two parts - combobox flag + invalid flags and then combobox only.

I feel the code might be simplified, so any suggestions are very appreciated.

Reply:
The unit test checks if the "comb" mode for text widget annotations is only set when the field flags are valid. The "comb" mode splits the text field into equal regions of a fixed amount of characters. For example, a field to enter a serial number that consists of three blocks of four characters will be rendered as as 1234 | 5678 | 9012, equally spaced in the text field. It checks this by setting all invalid field flags and removing them one by one in a loop until all invalid field flags are gone. The "comb" mode should only be set if the invalid field flags are all removed, which is what we assert.

However, since you're only making the annotation API asynchonous, no other changes should be necessary for this unit test as far as I can tell. In other words, I think the following should work for your purposes:

```javascript
    it('should only accept comb fields when the flags are valid', function() {
      let invalidFieldFlags = [
        AnnotationFieldFlag.MULTILINE,
        AnnotationFieldFlag.PASSWORD,
        AnnotationFieldFlag.FILESELECT,
      ];

      // Start with all invalid flags set and remove them one by one.
      // The field may only use combs when all invalid flags are unset.
      let flags = AnnotationFieldFlag.COMB + AnnotationFieldFlag.MULTILINE +
                 AnnotationFieldFlag.PASSWORD + AnnotationFieldFlag.FILESELECT;

      for (let i = 0, ii = invalidFieldFlags.length; i <= ii; i++) {
        textWidgetDict.set('MaxLen', 20);
        textWidgetDict.set('Ff', flags);

        let textWidgetRef = new Ref(93, 0);
        let xref = new XRefMock([
          { ref: textWidgetRef, data: textWidgetDict, }
        ]);

        AnnotationFactory.create(xref, textWidgetRef, pdfManagerMock,
            idFactoryMock).then((annotation) => {
          let data = annotation.data;
          expect(data.annotationType).toEqual(AnnotationType.WIDGET);

          let valid = (invalidFieldFlags.length === 0);
          expect(data.comb).toEqual(valid);
        });

        // Remove the last invalid flag for the next iteration.
        if (!valid) {
          flags -= invalidFieldFlags.splice(-1, 1);
        }
      }
    });
```

Hopefully this helps to understand the test a bit better. If not, or the test turns out to be more complicated than I thought here, feel free to let me know and I'll try to provide more explanation or look into any difficulties!

Reply:
The problem with this code (and as far as I remember it was exactly what I tried to address by splitting loop into 2 parts) is the 'valid' variable. It's initiated asynchronously and as result you can't use it in the condition 'if (!valid)' directly, unit test fails with the message "ReferenceError: valid is not defined". Plus if invalidFieldFlag is spliced synchronously I'm not sure if you can rely on it in the asynchronous l-function, it might cause a race condition.

Reply:
Yes, I now understand why you made that choice. For now I don't really see a better way, so let's keep it like that. Thank you for explaining this!